workflow/controller/operator.go

```diff
@@ -802,7 +802,7 @@ func (woc *wfOperationCtx) persistUpdates(ctx context.Context) {
 
 func (woc *wfOperationCtx) checkTaskResultsCompleted() bool {
 	taskResultsCompleted := woc.wf.Status.GetTaskResultsCompleted()
-	if len(taskResultsCompleted) == 0 {
+	if len(taskResultsCompleted) == 0 && woc.wf.Status.Outputs != nil {
 		return false
 	}
```

Comment:
Sorry, do you mind if we first go back to the original code and talk about the case you were trying to take care of with `if len(taskResultsCompleted) == 0 {`?

Reply:
BTW, maybe we should rename `taskResultsCompleted` to something like `taskResultCompletionStatus` since it sounds like we're talking about the number completed here but we're not.?

Reply:
As far as I can tell, we're looking at the case in which no taskresults have ever even been initialized as incomplete, right?

Reply:
Yeah, we could probably do with a rename.

Yes, we're looking at the case where no task results have been initialized as incomplete yet.

Reply:
Is this the case we're concerned about, or is it more generally speaking the case in which a task result would be initialized in the future and hasn't been yet? As far as I can tell, it's once a Pod exists that a new task result will get added? So, wouldn't it be the same case if say this map has one entry and it's completed, but another task result simply hasn't been initialized yet?

Reply:
💩 Yes, you're right. We should be handling that. I'm going to think on this a bit more & try to come up with something better.

Reply:
I've flipped the way I'm looking at this and reformulated the check as `checkTaskResultsInProgress`. When a task result hasn't completed yet, it is added to a `TaskResultsInProgress` map. Once it is completed, it is removed from the map.

```
func (woc *wfOperationCtx) checkTaskResultsInProgress() bool {
	taskResultsInProgress := woc.wf.Status.GetTaskResultsInProgress()
	if len(taskResultsInProgress) != 0 {
		return true
	}
	return false
}
```

In all three cases, `!woc.checkTaskResultsInProgress()` is called. All tests seem to be passing & the code is much easier to understand. I'm re-running the test script from my last PR to ensure the garbage collection race condition doesn't exist. https://github.com/argoproj/argo-workflows/pull/11947#issuecomment-1805074067.

Assuming all the tests pass locally, I'll push and you can review @juliev0

Reply:
@juliev0 The case of "no task results have been initialized as incomplete yet" doesn't actually seem to be an issue based on my testing.